File: app/src/main/java/app/kamy/saatApp/infrastructure/local/LocalKhgtCalendar.kt
package app.kamy.saatApp.infrastructure.local

import android.content.Context
import app.kamy.saatApp.domain.model.KhgtCalendarResponse
import app.kamy.saatApp.domain.model.KhgtDay
import app.kamy.saatApp.domain.model.KhgtMonth
import app.kamy.saatApp.domain.model.KhgtSpecialDay
import app.kamy.saatApp.domain.model.KhgtTodayInfo
import dagger.hilt.android.qualifiers.ApplicationContext
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import java.util.zip.GZIPInputStream
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

import app.kamy.saatApp.domain.model.RamadanDayInfo

@Singleton
class LocalKhgtCalendar @Inject constructor(
    @ApplicationContext private val context: Context,
    private val json: Json
) {
    @Volatile
    private var cachedYears: MutableMap<Int, KhgtCalendarResponse> = mutableMapOf()

    suspend fun todayInfo(): KhgtTodayInfo? = infoForDate(Calendar.getInstance())

    suspend fun ramadanInfo(
        date: Calendar = Calendar.getInstance(),
        imsakTime: String? = null,
        iftarTime: String? = null
    ): RamadanDayInfo? = withContext(Dispatchers.IO) {
        val gregorian = SimpleDateFormat("dd-MMM-yyyy", Locale.ENGLISH).format(date.time)
        val hijriYears = listOf(
            currentHijriYearGuess(date) - 1,
            currentHijriYearGuess(date),
            currentHijriYearGuess(date) + 1
        )
        for (year in hijriYears) {
            val calendar = loadYear(year) ?: continue
            val months = calendar.data.orEmpty()
            for (month in months) {
                if (month.name.equals("Ramadan", ignoreCase = true)) {
                    val days = month.days.orEmpty()
                    val dayIndex = days.indexOfFirst { it.masehi == gregorian }
                    if (dayIndex != -1) {
                        val day = days[dayIndex]
                        val dayNumber = parseArabicHijriDay(day.hijri) ?: (dayIndex + 1)
                        val totalDays = days.size.coerceAtLeast(29)
                        val hijriYear = month.year ?: year
                        return@withContext RamadanDayInfo(
                            isRamadan = true,
                            dayNumber = dayNumber,
                            totalDays = totalDays,
                            hijriYear = hijriYear,
                            hijriLabel = "$dayNumber Ramadan $hijriYear H",
                            imsakTime = imsakTime,
                            iftarTime = iftarTime
                        )
                    }
                }
            }
        }
        null
    }

    suspend fun infoForDate(date: Calendar): KhgtTodayInfo? = withContext(Dispatchers.IO) {
        val gregorian = SimpleDateFormat("dd-MMM-yyyy", Locale.ENGLISH).format(date.time)
        val gregorianShort = SimpleDateFormat("dd MMM yyyy", Locale.getDefault()).format(date.time)

        for (year in listOf(currentHijriYearGuess(date) - 1, currentHijriYearGuess(date), currentHijriYearGuess(date) + 1)) {
            val calendar = loadYear(year) ?: continue
            val day = findDayByGregorian(calendar.data.orEmpty(), gregorian) ?: continue
            val hijriLabel = buildString {
                append(day.hijri.orEmpty().trim())
                calendar.data?.firstOrNull { month ->
                    month.days?.any { it.masehi == gregorian } == true
                }?.name?.let { monthName ->
                    append(' ')
                    append(monthName)
                }
                append(' ')
                append(year)
            }.trim()
            val event = day.tooltip?.takeIf { it.isNotBlank() && day.isEvent == true }
            return@withContext KhgtTodayInfo(
                hijriLabel = hijriLabel,
                gregorianLabel = gregorianShort,
                pasaran = day.pasaran,
                eventTitle = event,
                isImportantDay = event != null
            )
        }
        null
    }

    suspend fun upcomingEvents(limit: Int = 5): List<Pair<String, String>> = withContext(Dispatchers.IO) {
        val today = Calendar.getInstance()
        val formatter = SimpleDateFormat("dd-MMM-yyyy", Locale.ENGLISH)
        val events = mutableListOf<Pair<Long, Pair<String, String>>>()

        for (yearOffset in -1..1) {
            val year = currentHijriYearGuess(today) + yearOffset
            val calendar = loadYear(year) ?: continue
            calendar.data.orEmpty().forEach { month ->
                month.days.orEmpty().forEach { day ->
                    val title = day.tooltip?.takeIf { it.isNotBlank() && day.isEvent == true } ?: return@forEach
                    val masehi = day.masehi ?: return@forEach
                    val date = runCatching { formatter.parse(masehi)?.time }.getOrNull() ?: return@forEach
                    if (date >= today.timeInMillis - 86_400_000L) {
                        events += date to (day.masehiShort.orEmpty() to title)
                    }
                }
            }
            calendar.specialDays.orEmpty().forEach { special ->
                val title = special.keterangan?.takeIf { it.isNotBlank() } ?: return@forEach
                val masehi = special.tanggalMasehi ?: return@forEach
                val date = runCatching { formatter.parse(masehi)?.time }.getOrNull() ?: return@forEach
                if (date >= today.timeInMillis - 86_400_000L) {
                    events += date to (masehi to title)
                }
            }
        }

        events.sortedBy { it.first }
            .map { it.second }
            .distinct()
            .take(limit)
    }

    suspend fun monthForToday(): KhgtMonth? = withContext(Dispatchers.IO) {
        val today = Calendar.getInstance()
        val gregorian = SimpleDateFormat("dd-MMM-yyyy", Locale.ENGLISH).format(today.time)
        for (year in listOf(currentHijriYearGuess(today) - 1, currentHijriYearGuess(today), currentHijriYearGuess(today) + 1)) {
            val calendar = loadYear(year) ?: continue
            calendar.data.orEmpty().forEach { month ->
                if (month.days?.any { it.masehi == gregorian } == true) return@withContext month
            }
        }
        null
    }

    private fun parseArabicHijriDay(arabic: String?): Int? {
        if (arabic.isNullOrBlank()) return null
        val western = arabic.map { ch ->
            when (ch) {
                '٠' -> '0'
                '١' -> '1'
                '٢' -> '2'
                '٣' -> '3'
                '٤' -> '4'
                '٥' -> '5'
                '٦' -> '6'
                '٧' -> '7'
                '٨' -> '8'
                '٩' -> '9'
                else -> ch
            }
        }.joinToString("")
        return western.filter { it.isDigit() }.toIntOrNull()
    }

    private fun loadYear(hijriYear: Int): KhgtCalendarResponse? {
        cachedYears[hijriYear]?.let { return it }
        val asset = "hijri/khgt_$hijriYear.json.gz"
        val plain = "hijri/khgt_$hijriYear.json"
        val text = when {
            assetExists(asset) -> readGzipAsset(asset)
            assetExists(plain) -> readAsset(plain)
            else -> return null
        }
        val parsed = json.decodeFromString(KhgtCalendarResponse.serializer(), text)
        cachedYears[hijriYear] = parsed
        return parsed
    }

    private fun findDayByGregorian(months: List<KhgtMonth>, gregorian: String): KhgtDay? =
        months.asSequence()
            .flatMap { it.days.orEmpty().asSequence() }
            .firstOrNull { it.masehi == gregorian }

    private fun currentHijriYearGuess(today: Calendar): Int {
        val localDate = today.toInstant()
            .atZone(today.timeZone.toZoneId())
            .toLocalDate()
        return java.time.chrono.HijrahDate.from(localDate)
            .get(java.time.temporal.ChronoField.YEAR_OF_ERA)
    }

    private fun assetExists(path: String): Boolean =
        runCatching { context.assets.open(path).close(); true }.getOrDefault(false)

    private fun readAsset(path: String): String =
        context.assets.open(path).bufferedReader().use { it.readText() }

    private fun readGzipAsset(path: String): String =
        context.assets.open(path).use { input ->
            GZIPInputStream(input).bufferedReader().use { it.readText() }
        }
}

File: iosApp/Saat/Infrastructure/Persistence/Local/LocalKhgtCalendar.swift
//
//  LocalKhgtCalendar.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

struct KhgtCalendarResponse: Codable, Sendable {
    let success: Bool?
    let message: String?
    let data: [KhgtMonth]?
    let specialDays: [KhgtSpecialDay]?

    enum CodingKeys: String, CodingKey {
        case success, message, data
        case specialDays = "special_days"
    }
}

struct KhgtMonth: Codable, Sendable {
    let name: String?
    let year: Int?
    let masehiRange: String?
    let days: [KhgtDay]?

    enum CodingKeys: String, CodingKey {
        case name, year
        case masehiRange = "masehi_range"
        case days
    }
}

struct KhgtDay: Codable, Sendable {
    let masehi: String?
    let masehiShort: String?
    let hijri: String?
    let pasaran: String?
    let tooltip: String?
    let isEvent: Bool?

    enum CodingKeys: String, CodingKey {
        case masehi
        case masehiShort = "masehi_short"
        case hijri, pasaran, tooltip
        case isEvent = "is_event"
    }
}

struct KhgtSpecialDay: Codable, Sendable {
    let tanggalHijri: String?
    let tanggalMasehi: String?
    let keterangan: String?

    enum CodingKeys: String, CodingKey {
        case tanggalHijri = "tanggal_hijri"
        case tanggalMasehi = "tanggal_masehi"
        case keterangan
    }
}

struct KhgtTodayInfo: Codable, Hashable, Sendable {
    let hijriLabel: String
    let gregorianLabel: String
    let pasaran: String?
    let eventTitle: String?
    let isImportantDay: Bool
}

internal final class LocalKhgtCalendar: Sendable {
    internal static let shared = LocalKhgtCalendar()

    private init() {}

    internal func todayInfo() -> KhgtTodayInfo? {
        return infoForDate(Date())
    }

    internal func infoForDate(_ date: Date) -> KhgtTodayInfo? {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US")
        formatter.dateFormat = "dd-MMM-yyyy"
        let gregorian = formatter.string(from: date)

        let displayFormatter = DateFormatter()
        displayFormatter.locale = Locale.current
        displayFormatter.dateFormat = "dd MMM yyyy"
        let gregorianShort = displayFormatter.string(from: date)

        // Guess Hijri year
        let hijriCalendar = Calendar(identifier: .islamicUmmAlQura)
        let hijriYear = hijriCalendar.component(.year, from: date)

        // Search in adjacent years if guess is slightly off due to moon sighting
        for year in [hijriYear, hijriYear + 1, hijriYear - 1] {
            guard let calResponse = loadYear(year) else { continue }
            guard let months = calResponse.data else { continue }

            for month in months {
                guard let days = month.days else { continue }
                if let day = days.first(where: { $0.masehi == gregorian }) {
                    let hijriLabel = "\(day.hijri ?? "") \(month.name ?? "") \(year)".trimmingCharacters(in: .whitespacesAndNewlines)
                    let event = (day.isEvent == true && !(day.tooltip ?? "").isEmpty) ? day.tooltip : nil
                    return KhgtTodayInfo(
                        hijriLabel: hijriLabel,
                        gregorianLabel: gregorianShort,
                        pasaran: day.pasaran,
                        eventTitle: event,
                        isImportantDay: event != nil
                    )
                }
            }
        }
        return nil
    }

    internal func monthForToday() -> KhgtMonth? {
        let date = Date()
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US")
        formatter.dateFormat = "dd-MMM-yyyy"
        let gregorian = formatter.string(from: date)

        let hijriCalendar = Calendar(identifier: .islamicUmmAlQura)
        let hijriYear = hijriCalendar.component(.year, from: date)

        for year in [hijriYear, hijriYear + 1, hijriYear - 1] {
            guard let calResponse = loadYear(year) else { continue }
            guard let months = calResponse.data else { continue }

            for month in months {
                if month.days?.contains(where: { $0.masehi == gregorian }) == true {
                    return month
                }
            }
        }
        return nil
    }

    private func loadYear(_ hijriYear: Int) -> KhgtCalendarResponse? {
        let filename = "khgt_\(hijriYear)"
        // Support flat or subdirectory structures
        guard let url = Bundle.main.url(forResource: filename, withExtension: "json") ??
                        Bundle.main.url(forResource: filename, withExtension: "json", subdirectory: "hijri") else {
            return nil
        }
        do {
            let data = try Data(contentsOf: url)
            return try JSONDecoder().decode(KhgtCalendarResponse.self, from: data)
        } catch {
            print("Failed to decode Hijri calendar for \(hijriYear): \(error.localizedDescription)")
            return nil
        }
    }
}
